### DISCUSSION TITLE
How to store and restore key mappings using nvim API?

### DISCUSSION THREAD

**Discussion by shmerl** [neovim/neovim#29463] [Q&A] [answered]
I'm trying to come up with something that reads current mapping, replaces it temporarily and then restores the original when needed.

Looking at available functions, I was trying to use these:

`nvim_set_keymap` - to set a mapping.
`nvim_get_keymap` - to read existing mapping.

But the latter presents a lot more information than the former. Lets say I preserve an existing mapping fishing out the value from that big table by mode and and lhs. How do I restore it later when I need to? These two functions even though are misleadingly named as reversible have very different data formats and aren't really reversible semantically unless I'm missing something?

So what is the right way to implement the above?

If it's not possible without some manual steps now, it would be good to have a couple of reversible functions.

I.e. one that reads the mapping and one the sets it possibly using the result of the other one mostly directly.

**Comment by clason** [comment, marked as answer]
Use `vim.fn.mapset`.

**Reply by shmerl** [reply]
`vim.fn.mapset` should work with the entry from the table returned by `nvim_get_keymap`?

I'll try. It would have been nice if there were a couple of symmetric functions in the style of `nvim_set_keymap` though.

**Reply by clason** [reply]
Yes. No need for a new function if there already exists one; "nice if there were symmetry" is not a strong enough argument.

**Reply by shmerl** [reply]
May be, but the naming kind of suggests such relation, which is a bit misleading I suppose.

**Reply by clason** [reply]
And might as well use `vim.fn.maparg` if you want symmetry; for an example, see https://github.com/neovim/neovim/blob/be999e6a0e5b251b2b37500d06636d4167334c6e/runtime/lua/vim/snippet.lua#L257

**Reply by clason** [reply]
> May be, but the naming kind of suggests such relation, which is a bit misleading I suppose.

Not beyond setting and getting mappings, which they do. Programmatically saving and restoring mappings is a different thing, for which an API already exists (`map{arg,set}`).

**Reply by shmerl** [reply]
Yeah, I know that's close to vimscript style setting. I was trying to use more Lua style which nvim_set_keymap sort of is doing.

**Reply by clason** [reply]
Why? `vim.fn` is part of the Lua API the same way as `vim.api`.

**Reply by shmerl** [reply]
Not sure, that's how it felt. vim.fn is mostly a verbatim wrapper over existing vimscript API. nvim_set_keymap felt more like regular Lua.

**Reply by clason** [reply]
Both are verbatim wrappers around C functions. Just because one came from Vim and the other was written later in Neovim with a different naming convention (and stronger API contract for external GUIs) doesn't make one better than the other.

**Reply by shmerl** [reply]
As long as it works. I'll try using `nvim_set_keymap` for setting custom stuff since it looks nicer, but I'll use your suggestion for reading / restoring the current one.

**Reply by shmerl** [reply]
By the way, `vim.keymap.set` looks pretty close in functionality to `vim.api.nvim_set_keymap`. Is there some recommended context when to use one or another and why are there are two in general?

**Reply by shmerl** [reply]
If anyone needs, found some details here:

https://github.com/neovim/neovim/commit/6d41f65aa45f10a93ad476db01413abaac21f27d

Basically, `vim.api.nvim_set_keymap` is the API function that has to preserve backward compatibility and `vim.keymap.set` is a neater version that doesn't.
